### PR TITLE
[TACHYON-590] Implement the wildcard support for shell cmds with 2 arguments

### PR DESCRIPTION
**Pull request by tianyin**
This is the last batch of TACHYON-590. It implements the wildcard support for shell cmds with 2 arguments, namely copyFromLocal and copyToLocal.

One behavior change is that this PR implements the support for copy directory using 'copyToLocal' (the original cmd only supports copy a file to local filesystem).

### CODE REVIEW

File: shell/src/main/java/tachyon/shell/TFsShell.java [hunk 1]
```diff
@@ -162,25 +162,58 @@ private int loadPath(TachyonFS tachyonClient, TachyonURI filePath) throws IOExce
...
-  public int copyFromLocal(String[] argv) throws IOException {
-    String srcPath = argv[1];
-    TachyonURI dstPath = new TachyonURI(argv[2]);
-    File src = new File(srcPath);
-    if (!src.exists()) {
-      System.out.println("Local path " + srcPath + " does not exist.");
+  public int copyFromLocalWildcard(List<File> srcFiles, TachyonURI dstPath) throws IOException {
+    TachyonFS tachyonClient = createFS(dstPath);
+    TachyonFile dstFile = tachyonClient.getFile(dstPath);
+    if (dstFile != null && dstFile.isDirectory() == false) {
```
Comment (uronce-cc) [inline shell/src/main/java/tachyon/shell/TFsShell.java]: `!dstFile.isDirectory()` is more common in codebase.

File: shell/src/main/java/tachyon/shell/TFsShell.java [hunk 2]
```diff
@@ -162,25 +162,58 @@ private int loadPath(TachyonFS tachyonClient, TachyonURI filePath) throws IOExce
...
+      System.out.println("The destination cannot be an existent file when the src contains " 
+          + "wildcards.");
       return -1;
     }
+    if (dstFile == null) {
+      if (tachyonClient.mkdirs(dstPath, true) == false) {
```
Comment (uronce-cc) [inline shell/src/main/java/tachyon/shell/TFsShell.java]: `!tachyonClient.mkdirs`

File: shell/src/main/java/tachyon/shell/TFsShell.java [hunk 3]
```diff
@@ -224,37 +257,123 @@ private int copyPath(File src, TachyonFS tachyonClient, TachyonURI dstPath) thro
...
+   * @param dstFile The destination directory in the local filesystem 
    * @return 0 if command is successful, -1 if an error occurred.
    * @throws IOException
    */
-  public int copyToLocal(String[] argv) throws IOException {
-    TachyonURI srcPath = new TachyonURI(argv[1]);
-    String dstPath = argv[2];
-    File dst = new File(dstPath);
+  public int copyWildcardToLocal(List<TachyonURI> srcPaths, File dstFile) throws IOException {
+    if (dstFile.exists() && dstFile.isDirectory() == false) {
```
Comment (uronce-cc) [inline shell/src/main/java/tachyon/shell/TFsShell.java]: `! dstFile.isDirectory()`

**Comment by uronce-cc** [comment]
NIT: `! boolean` is more common in the codebase than `boolean == false` where `boolean` refers to a `boolean` variable.

**Comment by uronce-cc** [comment]
LGTM overall!

**Comment by tianyin** [comment]
@chang, I'm in eu these days and cannot access internet easily. I'm going
back home next week and will definitely address your comments. Thank you so
much, as always! I've many bad habits and fortunately many of them are
correct by you gradually. <3

On Thursday, September 3, 2015, chang cheng <notifications@github.com>
wrote:

> LGTM overall!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/1418#issuecomment-137357032>.
>


-- 
Tianyin XU,
http://cseweb.ucsd.edu/~tixu/

File: shell/src/main/java/tachyon/shell/TFsShell.java [hunk 4]
```diff
@@ -162,25 +162,58 @@ private int loadPath(TachyonFS tachyonClient, TachyonURI filePath) throws IOExce
...
+        return -1;
+      } else {
+        System.out.println("Create directory: " + dstPath);
+      }
+    }
+    int exitCode = 0;
+    for (File srcFile : srcFiles) {
+      try {
+        exitCode |= copyFromLocal(srcFile, 
+            new TachyonURI(dstPath.getPath() + TachyonURI.SEPARATOR + srcFile.getName()));
```
Comment (apc999) [inline shell/src/main/java/tachyon/shell/TFsShell.java]: better to use `tachyon.util.io.PathUtils#concatPath` to construct the path string

File: shell/src/main/java/tachyon/shell/TFsShell.java [hunk 5]
```diff
@@ -224,37 +257,123 @@ private int copyPath(File src, TachyonFS tachyonClient, TachyonURI dstPath) thro
     }
     return 0;
   }
-
+
```
Comment (apc999) [inline shell/src/main/java/tachyon/shell/TFsShell.java]: unnecessary space in line 260

File: shell/src/main/java/tachyon/shell/TFsShell.java [hunk 6]
```diff
@@ -162,25 +162,58 @@ private int loadPath(TachyonFS tachyonClient, TachyonURI filePath) throws IOExce
...
+    }
+    return exitCode;
+  }
+  
+  /**
+   * Copies a file or directory specified by srcPath from the local filesystem to dstPath in 
+   * the Tachyon filesystem space. Will
+   * fail if the path given already exists in the filesystem.
+   *
+   * @param srcPath The path of the source file in the local filesystem
```
Comment (apc999) [inline shell/src/main/java/tachyon/shell/TFsShell.java]: javadoc is out-of-date---no `srcPath` but `srcFile`

**Comment by apc999** [comment]
looks in general good. Thanks for the patch!

**Comment by uronce-cc** [comment]
+1 LGTM!

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 7]
```diff
@@ -205,31 +205,42 @@ public void copyFromLocalFileToDstPathTest() throws IOException {
...
-    File testFile = new File(mLocalTachyonCluster.getTachyonHome() + "/testFile");
+    fileReadTest("/testFile", 10);
+  }
+  
+  @Test
+  public void copyToLocalDirTest() throws IOException {
+    TFsShellUtilsTest.resetTachyonFileHierarchy(mTfs, WriteType.MUST_CACHE);
+    int ret = mFsShell.run(new String[] {"copyToLocal", "/testWildCards/",
+        mLocalTachyonCluster.getTachyonHome() + "/testDir"});
+    Assert.assertEquals(ret, 0);
```
Comment (apc999) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: In `assertEquals`, we typically put the expected value as the first parameter and the actual test return as the second parameter, so the error message will be shown correctly.

Same comment apply to other occurrences of `assertEquals` in this PR.

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 8]
```diff
@@ -205,31 +205,42 @@ public void copyFromLocalFileToDstPathTest() throws IOException {
...
+    fileReadTest("/testDir/foo/foobar1", 10);
+    fileReadTest("/testDir/foo/foobar2", 20);
+    fileReadTest("/testDir/bar/foobar3", 30);
+    fileReadTest("/testDir/foobar4", 40);
+  }
+  
+  private void fileReadTest(String fileName, int size) throws IOException {
+    File testFile = new File(mLocalTachyonCluster.getTachyonHome() + "/" + fileName);
```
Comment (apc999) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: use `concatPath` please for better portability.

File: shell/src/test/java/tachyon/shell/TFsShellTest.java [hunk 9]
```diff
@@ -848,4 +859,97 @@ public void loadWildcardTest() throws IOException {
...
+  @Test
+  public void copyFromLocalWildcardTest() throws IOException {
+    TFsShellUtilsTest.resetLocalFileHierarchy(mLocalTachyonCluster);
+    int ret = mFsShell.run(new String[] {"copyFromLocal", 
+        mLocalTachyonCluster.getTachyonHome() + "/testWildCards/*/foo*", "/testDir"});
+    Assert.assertEquals(ret, 0);
+    Assert.assertNotNull(mTfs.getFile(new TachyonURI("/testDir/foobar1")));
+    Assert.assertNotNull(mTfs.getFile(new TachyonURI("/testDir/foobar2")));
+    Assert.assertNotNull(mTfs.getFile(new TachyonURI("/testDir/foobar3")));
+  }
```
Comment (apc999) [inline shell/src/test/java/tachyon/shell/TFsShellTest.java]: `assertNull` on "/testDir/foobar4" possibly ?

**Comment by apc999** [comment]
Thanks for the patch! I uploaded a bunch minor comments regarding unit tests.

**Comment by tianyin** [comment]
@apc999 thanks, Bin!

**Comment by apc999** [comment]
Thanks for putting it together, @tianyin 
+1 LGTM

**Pull request merged**